lib/Adapter/TolerantParser/Updater/ClassLikeUpdater.php

```diff
@@ -0,0 +1,168 @@
+<?php
+
+namespace Phpactor\CodeBuilder\Adapter\TolerantParser\Updater;
+
+use Microsoft\PhpParser\Node;
+use Microsoft\PhpParser\Node\ClassConstDeclaration;
+use Microsoft\PhpParser\Node\Expression\AssignmentExpression;
+use Microsoft\PhpParser\Node\Expression\Variable;
+use Microsoft\PhpParser\Node\MethodDeclaration;
+use Microsoft\PhpParser\Node\PropertyDeclaration;
+use Microsoft\PhpParser\Node\StatementNode;
+use Phpactor\CodeBuilder\Adapter\TolerantParser\Edits;
+use Phpactor\CodeBuilder\Domain\Prototype\ClassLikePrototype;
+use Phpactor\CodeBuilder\Domain\Prototype\Type;
+use Phpactor\CodeBuilder\Domain\Renderer;
+
+abstract class ClassLikeUpdater
+{
+    /**
+     * @var Renderer
+     */
+    protected $renderer;
+
+    /**
+     * @var MethodUpdater
+     */
+    protected $methodUpdater;
+
+    public function __construct(Renderer $renderer)
+    {
+        $this->renderer = $renderer;
+        $this->methodUpdater = new ClassMethodUpdater($renderer);
+    }
+
+    protected function resolvePropertyName(Node $property)
+    {
+        if ($property instanceof Variable) {
+            return $property->getName();
+        }
+
+        if ($property instanceof AssignmentExpression) {
+            return $this->resolvePropertyName($property->leftOperand);
+        }
+
+        throw new \InvalidArgumentException(sprintf(
+            'Do not know how to resolve property element of type "%s"',
+            get_class($property)
+        ));
+    }
+
+    protected function updateConstants(Edits $edits, ClassLikePrototype $classPrototype, StatementNode $classNode)
+    {
+        if (count($classPrototype->constants()) === 0) {
+            return;
+        }
+
+        switch ($classNode->getNodeKindName()) {
+            case 'ClassDeclaration':
+                $lastConstant = $classNode->classMembers->openBrace;
+                $memberDeclarations = $classNode->classMembers->classMemberDeclarations;
+                break;
+            case 'TraitDeclaration':
+                $lastConstant = $classNode->traitMembers->openBrace;
+                $memberDeclarations = $classNode->traitMembers->traitMemberDeclarations;
```

Comment:
I guess maybe we can just add an abstract method `memberDeclarations(): array`? which returns `Node[]` ?

Reply:
That's the most likely option, along with something similar for the `$lastConstant` definition.

That said, I completely forgot traits don't actually support constants. With that in mind, I'll probably move `updateConstants()` back to the concrete `ClassUpdater` later today.

Hopefully nothing breaks in the process, but it will at least cut out one of the two switches.